src/Handler/ProcessRefundHandler.php

```diff
@@ -0,0 +1,196 @@
+<?php
+
+namespace App\Handler;
+
+use App\Entity\MiraklRefund;
+use App\Entity\StripeTransfer;
+use App\Message\ProcessRefundMessage;
+use App\Repository\StripeTransferRepository;
+use App\Repository\MiraklRefundRepository;
+use App\Utils\MiraklClient;
+use App\Utils\StripeProxy;
+use Psr\Log\LoggerAwareInterface;
+use Psr\Log\LoggerAwareTrait;
+use Stripe\Exception\ApiErrorException;
+use Symfony\Component\Messenger\Handler\MessageHandlerInterface;
+
+class ProcessRefundHandler implements MessageHandlerInterface, LoggerAwareInterface
+{
+    use LoggerAwareTrait;
+
+    /**
+     * @var MiraklClient
+     */
+    private $miraklClient;
+
+    /**
+     * @var StripeProxy
+     */
+    private $stripeProxy;
+
+    /**
+     * @var StripeTransferRepository
+     */
+    private $stripeTransferRepository;
+
+    /**
+     * @var MiraklRefundRepository
+     */
+    private $miraklRefundRepository;
+
+    public function __construct(
+        MiraklClient $miraklClient,
+        StripeProxy $stripeProxy,
+        StripeTransferRepository $stripeTransferRepository,
+        MiraklRefundRepository $miraklRefundRepository
+    ) {
+        $this->miraklClient = $miraklClient;
+        $this->stripeProxy = $stripeProxy;
+        $this->stripeTransferRepository = $stripeTransferRepository;
+        $this->miraklRefundRepository = $miraklRefundRepository;
+    }
+
+    public function __invoke(ProcessRefundMessage $message)
+    {
+        $refund = $this->miraklRefundRepository->findOneBy([
+            'miraklRefundId' => $message->geMiraklRefundId(),
+        ]);
+
+        if (MiraklRefund::REFUND_CREATED == $refund->getStatus()) {
+            // Already processed nothing todo
+            return;
+        }
+
+        $transfer = $this->stripeTransferRepository->findOneBy([
+            'miraklId' => $refund->getMiraklOrderId(),
+            'status' => StripeTransfer::TRANSFER_CREATED,
+            'type' => StripeTransfer::TRANSFER_ORDER,
+        ]);
+
+        if (is_null($transfer)) {
+            $failedReason = sprintf('Mirakl refund id: %s and orderId: %s has no stripe transfer in connector', $refund->getMiraklRefundId(), $refund->getMiraklOrderId());
+            $refund
+                ->setStatus(MiraklRefund::REFUND_FAILED)
+                ->setFailedReason($failedReason);
+            $this->miraklRefundRepository->persistAndFlush($refund);
+
+            return;
+        }
+
+        try {
+            // Refund and transfer back in Stripe
+            $this->processRefundInStripe($refund, $transfer);
+            $this->processReversalInStripe($refund, $transfer);
+            // Valid refund on mirakl PA02
+            $this->validateRefundInMirakl($refund);
```

Comment:
If I understand correctly, you validate the refund on mirakl only if both refund + reverse transfer worked on Stripe. 
 If the seller already payed out, the reversal could fail, but the refund may be ok.

If the refund worked, the client will see the money, but the operator will not see the refund as successful in the interface.
Could it be a problem? The client could argue that they never received the refund.

Maybe we should validate the refund as soon as the refund is ok on Stripe, and warn the operator in case the reversal fails? @youssef-stripe what do you think?

Reply:
I agree with you Kevin. The refund should be marked as valid in Mirakl if `processRefundInStripe` succeeds. Then if processReversalInStripe fails, it should be handled the same as a Transfer failure: reverse transfer is later retried and eventually an alert is sent to the administrator.

The reason is that you wouldn't reverse the transfer if the refund fails (which it can) while the opposite is not true as you'd still want to refund the customer even if the transfer cannot be reversed immediately.

What do you think Alexandra?

Reply:
Hi, 
Yes, it will make more sense to confirm the refund in Mirakl once the refund has been done to the customer. 
In this case, I will change the way I handle transfer reversal and do it in the `ProcessTransferHandler`.